File: Python/swapPairs.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: swapPairs.py
@Function: 两两交换链表中的节点
@Link: https://leetcode-cn.com/problems/swap-nodes-in-pairs/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-30
"""


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        dummyHead = ListNode(0)
        dummyHead.next = head
        temp = dummyHead
        while temp.next and temp.next.next:
            node1 = temp.next
            node2 = temp.next.next
            temp.next = node2
            node1.next = node2.next
            node2.next = node1
            temp = node1
        return dummyHead.next


# --------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass

File: C++/swapPairs.cpp
/**
 * @File    : swapPairs.cpp
 * @Brief   : 两两交换链表中的节点
 * @Link    : https://leetcode-cn.com/problems/swap-nodes-in-pairs/
 * @Author  : Wei Li
 * @Date    : 2021-07-30
*/

#include <iostream>

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// 利用指针进行交换操作
class Solution
{
public:
    ListNode *swapPairs(ListNode *head)
    {
        ListNode *p = head, *s;
        if (p && p->next)
        {
            s = p->next;
            p->next = s->next;
            s->next = p;
            head = s;
            while (p->next && p->next->next)
            {
                s = p->next->next;
                p->next->next = s->next;
                s->next = p->next;
                p->next = s;
                p = s->next;
            }
        }
        return head;
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}
